README.md
<div align="center">
  <img width="471" style="max-width:100%;" src="https://github.com/futurestudio/hapi-response-utilities/blob/master/media/hapi-response-utilities.png?raw=true" alt="hapi-response-utilities logo">
  <br/>
  <br/>
  <p>
    hapi response decorations to conveniently compose responses
  </p>
  <br/>
  <p>
    <a href="#installation"><strong>Installation</strong></a> ·
    <a href="#usage"><strong>Usage</strong></a> ·
    <a href="#methods"><strong>Methods</strong></a>
  </p>
  <br/>
  <br/>
  <p>

[![Build Status](https://travis-ci.org/futurestudio/hapi-response-utilities.svg?branch=master)](https://travis-ci.org/futurestudio/hapi-response-utilities) [![Greenkeeper badge](https://badges.greenkeeper.io/futurestudio/hapi-response-utilities.svg)](https://greenkeeper.io/)
  <a href="https://snyk.io/test/github/futurestudio/hapi-response-utilities"><img src="https://snyk.io/test/github/futurestudio/hapi-response-utilities/badge.svg" alt="Known Vulnerabilities" data-canonical-src="https://snyk.io/test/github/futurestudio/hapi-response-utilities" style="max-width:100%;"></a>
    <a href="https://www.npmjs.com/package/hapi-response-utilities"><img src="https://img.shields.io/npm/v/hapi-response-utilities.svg" alt="hapi-response-utilities Version"></a>
    <a href="https://www.npmjs.com/package/hapi-response-utilities"><img src="https://img.shields.io/npm/dm/hapi-response-utilities.svg" alt="Monthly downloads"></a>
  </p>
  <p>
    <em>Follow <a href="http://twitter.com/marcuspoehls">@marcuspoehls</a> for updates!</em>
  </p>
</div>

------

<p align="center"><sup>Development of this hapi plugin is supported by <a href="https://futurestud.io">Future Studio University 🚀</a></sup>
<br><b>
Join the <a href="https://futurestud.io/university">Future Studio University and Skyrocket in Node.js</a></b>
</p>

------


## Introduction
A hapi plugin that decorates the response toolkit `h` with methods to conveniently compose responses.


## Requirements
> **hapi v19 (or later)** and **Node.js v12 (or newer)**

This plugin requires **hapi v19** (or later) and **Node.js v12 or newer**.


### Compatibility
| Major Release | [hapi.js](https://github.com/hapijs/hapi) version | Node.js version |
| --- | --- | --- |
| `v2` | `>=17 hapi` | `>=12` |
| `v1` | `>=17 hapi` | `>=8` |


## Installation
Add `hapi-response-utilities` as a dependency to your project:

```bash
npm i hapi-response-utilities
```


## Usage
Register `hapi-response-utilities` to your hapi server and that's it :)

```js
await server.register({
  plugin: require('hapi-response-utilities')
})

// went smooth like hot chocolate :)
```


## Methods
An overview of available hapi `toolkit` decorations.


#### h.pdf(pdf-content, filename)
Creates a PDF response including the PDF related content type and HTTP headers allowing a custom filename.

The `filename` defaults to `download`.

```js
handler: async (_, h) => {
  const content = await createPdfContent()
  return h.pdf(content, 'filename.pdf')
}
```


#### h.status(code)
Shortcut to respond with just an HTTP status code.

```js
handler: (_, h) => {
  return h.status(204)
}
```


#### h.cookie(key, value, options)
Sets a cookie for the given key-value-pair. It's a convenience method for `h.state`.

```js
handler: (_, h) => {
  return
    h.cookie('userId', '1')
     .cookie('username', 'Marcus')
     .continue
}
```


#### h.header(key, value, options)
Set a response header for the given key-value-pair. This method provides a unified interface to set response headers, no matter if the response is a hapi response object or a boom error instance:

```js
handler: (_, h) => {
  return h
    .header('content-type', 'text/html')
    .header('api-key', 'secret-api-key')
    .continue
}
```

The accepted [`options`](https://hapi.dev/api/?v=18.4.0#-responseheadername-value-options) are the same as provided by hapi itself.


#### h.headers()
Returns an object containing the response headers. This method provides a unified interface to set response headers, no matter if the response is a hapi response object or a boom error instance:

```js
handler: (_, h) => {
  const responseHeaders = h.headers()

  return h.continue
}
```


#### h.isView()
Determines whether the response is a rendered view:

```js
handler: (_, h) => {
  if (h.isView()) {
    // handle view response
  }

  return h.continue
}
```


## Feature Requests
Do you miss a feature? Please don’t hesitate to
[create an issue](https://github.com/futurestudio/hapi-response-utilities/issues) with a short description of your desired addition to this plugin.


## Links & Resources

- [hapi tutorial series](https://futurestud.io/tutorials/hapi-get-your-server-up-and-running) with 100+ tutorials


## Contributing

1.  Create a fork
2.  Create your feature branch: `git checkout -b my-feature`
3.  Commit your changes: `git commit -am 'Add some feature'`
4.  Push to the branch: `git push origin my-new-feature`
5.  Submit a pull request 🚀


## License

MIT © [Future Studio](https://futurestud.io)

---

> [futurestud.io](https://futurestud.io) &nbsp;&middot;&nbsp;
> GitHub [@futurestudio](https://github.com/futurestudio/) &nbsp;&middot;&nbsp;
> Twitter [@futurestud_io](https://twitter.com/futurestud_io)


CHANGELOG.md
# Changelog


## Version [2.0.0](https://github.com/futurestudio/hapi-response-utilities/compare/v1.6.0...v2.0.0) - 2020-01-10

### Breaking Changes
- require Node.js v12
  - this change aligns with the hapi ecosystem requiring Node.js v12 with the release of hapi 19


## Version [1.6.0](https://github.com/futurestudio/hapi-response-utilities/compare/v1.5.0...v1.6.0) - 2019-12-13

### Added
- `.isView()`: determine whether the response is a rendered view
- `.headers()`: provides a unified interface returning the response headers (no matter if hapi resonse or boom error)
- `.header(key, value, options)`: provides a unified interface to set response headers. Works for both, a hapi response object or a boom error

### Updated
- bump dependencies


## Version [1.5.0](https://github.com/futurestudio/hapi-response-utilities/compare/v1.4.1...v1.5.0) - 2019-10-17

### Added
- basic TypeScript declarations in `lib/index.d.ts`


## Version [1.4.1](https://github.com/futurestudio/hapi-response-utilities/compare/v1.4.0...v1.4.1) - 2019-10-12

### Updated
- bump dependencies


## Version [1.4.0](https://github.com/futurestudio/hapi-response-utilities/compare/v1.3.4...v1.4.0) - 2019-07-24

### Added
- `.redirectWithPayload(path)`: temporary redirect with HTTP status 307 (keeps the HTTP method)
- `.permanentRedirectWithPayload(path)`: permanent redirect with HTTP status 308 (keeps the HTTP method)

### Updated
- bump dependencies


## Version [1.3.4](https://github.com/futurestudio/hapi-response-utilities/compare/v1.3.3...v1.3.4) - 2019-04-24

### Updated
- bump dependencies
- update to hapi scoped dependencies


## Version [1.3.3](https://github.com/futurestudio/hapi-response-utilities/compare/v1.3.2...v1.3.3) - 2019-02-18

### Updated
- bump dependencies
- fix badges in Readme


## Version [1.3.2](https://github.com/futurestudio/hapi-response-utilities/compare/v1.3.1...v1.3.2) - 2019-01-26

### Updated
- Readme: rename GitHub references `futurestudio -> futurestudio`


## Version [1.3.1](https://github.com/futurestudio/hapi-response-utilities/compare/v1.3.0...v1.3.1) - 2019-01-22

### Updated
- bump dependencies
- test plugin for hapi 18


## [1.3.0](https://github.com/futurestudio/hapi-response-utilities/compare/v1.2.0...v1.3.0) - 2019-01-09

### Added
- `h.permanentRedirect(path)`: sends a redirect with HTTP status code for permanently moved `301`


## [1.2.0](https://github.com/futurestudio/hapi-response-utilities/compare/v1.1.0...v1.2.0) - 2018-11-14

### Added
- Shortcut methods that set the related status code:
  - `h.ok`: `200`
  - `h.created`: `201`
  - `h.accepted`: `202`
  - `h.nonAuthoritativeInformation`: `203`
  - `h.noContent`: `204`
  - `h.resetContent`: `205`
  - `h.partialContent`: `206`

## Updated
- bump devDependencies


## [1.1.0](https://github.com/futurestudio/hapi-response-utilities/compare/v1.0.1...v1.1.0) - 2018-11-12

### Updated
- `.cookie()` function returns the response toolkit instead of `undefined`
- Readme: add logo
- Readme: example code snippets
- Run tests for Node versions 8, 10, and 11
- Bump dependencies


## [1.0.1](https://github.com/futurestudio/hapi-response-utilities/compare/v1.0.0...v1.0.1) - 2018-08-21

### Updated
- Readme: quick navigation


## 1.0.0 - 2018-08-02

### Added
- `1.0.0` release 🚀 🎉


test/index.js
'use strict'

const Path = require('path')
const Lab = require('@hapi/lab')
const Hapi = require('@hapi/hapi')
const Boom = require('@hapi/boom')
const Vision = require('@hapi/vision')
const { expect } = require('@hapi/code')

let server

const { experiment, it, beforeEach } = (exports.lab = Lab.script())

experiment('hapi-response-utilities', () => {
  beforeEach(async () => {
    server = new Hapi.Server({
      routes: {
        validate: {
          failAction: (_, __, error) => {
            throw error
          }
        }
      }
    })

    await server.initialize()

    await server.register({
      plugin: require('../lib/index')
    })
  })

  it('status()', async () => {
    server.route({
      path: '/',
      method: 'GET',
      handler: (_, h) => {
        return h.status(204)
      }
    })

    const request = {
      url: '/',
      method: 'GET'
    }

    const response = await server.inject(request)
    expect(response.statusCode).to.equal(204)
  })

  it('cookie()', async () => {
    server.route({
      path: '/',
      method: 'GET',
      handler: (_, h) => {
        return h
          .cookie('userId', '1')
          .cookie('username', 'Marcus')
          .continue
      }
    })

    const request = {
      url: '/',
      method: 'GET'
    }

    const response = await server.inject(request)
    expect(response.statusCode).to.equal(204)

    const cookies = response.headers['set-cookie'].join(';').split(';')
    expect(cookies).to.include('userId=1')
    expect(cookies).to.include('username=Marcus')
  })

  it('pdf()', async () => {
    server.route({
      path: '/',
      method: 'GET',
      handler: (_, h) => {
        return h.pdf(null, 'hapi-response-utilities')
      }
    })

    const request = {
      url: '/',
      method: 'GET'
    }

    const response = await server.inject(request)
    expect(response.statusCode).to.equal(204)
    expect(response.headers['content-type']).to.equal('application/pdf')
    expect(response.headers['content-disposition']).to.contain('hapi-response-utilities')
    expect(response.payload).to.exist()
  })

  experiment('header()', () => {
    const testHeadersWith = async (method) => {
      const server = Hapi.server()

      await server.register({
        plugin: require('../lib/index')
      })

      server.route({
        method: 'GET',
        path: '/non-error',
        options: {
          handler: () => ({ success: true }),
          ext: {
            onPreResponse: { method }
          }
        }
      })

      server.route({
        method: 'GET',
        path: '/error',
        options: {
          handler: () => {
            throw Boom.unauthorized('Original message')
          },
          ext: {
            onPreResponse: { method }
          }
        }
      })

      return server
    }

    it('throws when response is not available', async () => {
      server.route({
        path: '/',
        method: 'GET',
        handler: (_, h) => {
          expect(() => h.header('key', 'value')).to.throw()

          return 'success'
        }
      })

      const response = await server.inject('/')
      expect(response.payload).to.equal('success')
    })

    it('sets headers without any options', async () => {
      const server = await testHeadersWith((_, h) => {
        return h
          .header('a', 'x')
          .header('b', 'x')
          .header('b', 'y')
          .continue
      })

      const { headers: errorHeaders } = await server.inject('/error')
      const { headers: nonErrorHeaders } = await server.inject('/non-error')

      expect(errorHeaders).to.contain({ a: 'x', b: 'y' })
      expect(nonErrorHeaders).to.contain({ a: 'x', b: 'y' })
    })

    it('does not set existing header when override is false', async () => {
      const server = await testHeadersWith((_, h) => {
        return h
          .header('a', 'x', { override: false })
          .header('a', 'y', { override: false })
          .continue
      })

      const { headers: errorHeaders } = await server.inject('/error')
      const { headers: nonErrorHeaders } = await server.inject('/non-error')

      expect(errorHeaders).to.contain({ a: 'x' })
      expect(nonErrorHeaders).to.contain({ a: 'x' })
    })

    it('appends to existing headers with separator', async () => {
      const server = await testHeadersWith((_, h) => {
        return h
          .header('a', 'x', { append: true })
          .header('A', 'y', { append: true })
          .header('b', 'x', { append: true, separator: ';' })
          .header('B', 'y', { append: true, separator: ';' })
          .continue
      })

      const { headers: errorHeaders } = await server.inject('/error')
      const { headers: nonErrorHeaders } = await server.inject('/non-error')

      expect(errorHeaders).to.contain({ a: 'x,y', b: 'x;y' })
      expect(nonErrorHeaders).to.contain({ a: 'x,y', b: 'x;y' })
    })

    it('handles special case for appending set-cookie', async () => {
      const server = await testHeadersWith((_, h) => {
        return h
          .header('set-cookie', 'a=x', { append: true })
          .header('set-cookie', 'b=x', { append: true })
          .header('set-cookie', 'b=y', { append: true })
          .continue
      })

      const { headers: errorHeaders } = await server.inject('/error')
      const { headers: nonErrorHeaders } = await server.inject('/non-error')

      expect(nonErrorHeaders).to.contain({ 'set-cookie': ['a=x', 'b=x', 'b=y'] })
      expect(errorHeaders).to.contain({ 'set-cookie': ['a=x', 'b=x', 'b=y'] })
    })

    it('prevents duplicates when appending when duplicate is false', async () => {
      const server = await testHeadersWith((_, h) => {
        return h
          .header('a', 'x', { append: true })
          .header('A', 'y', { append: true })
          .header('a', 'y', { append: true, duplicate: false })
          .header('b', 'x', { append: true, separator: ';' })
          .header('B', 'y', { append: true, separator: ';' })
          .header('b', 'y', { append: true, separator: ';', duplicate: false })
          .continue
      })

      const { headers: errorHeaders } = await server.inject('/error')
      const { headers: nonErrorHeaders } = await server.inject('/non-error')

      expect(errorHeaders).to.contain({ a: 'x,y', b: 'x;y' })
      expect(nonErrorHeaders).to.contain({ a: 'x,y', b: 'x;y' })
    })
  })

  experiment('headers()', () => {
    it('throws when response is not available', async () => {
      server.route([
        {
          path: '/error',
          method: 'GET',
          handler: (_, h) => {
            expect(() => h.headers()).to.throw()
            return 'error'
          }
        },
        {
          path: '/no-error',
          method: 'GET',
          options: {
            handler: (_, h) => {
              expect(() => h.headers()).to.throw()
              return 'error'
            },
            ext: {
              onPreResponse: {
                method: (_, h) => {
                  expect(() => h.headers).to.not.throw()
                  return 'no-error'
                }
              }
            }
          }
        }
      ])

      const { payload: errorPayload } = await server.inject('/error')
      const { payload: noErrorPayload } = await server.inject('/no-error')
      expect(errorPayload).to.equal('error')
      expect(noErrorPayload).to.equal('no-error')
    })

    it('gets headers values from a non-error response', async () => {
      let headers

      server.route({
        path: '/',
        method: 'GET',
        options: {
          handler: () => {
            return 'success'
          },
          ext: {
            onPreResponse: {
              method: (request, h) => {
                request.response.header('a', 'x')
                request.response.header('b', 'y')

                headers = h.headers()

                return h.continue
              }
            }
          }
        }
      })

      const { headers: responseHeaders } = await server.inject('/')
      expect(headers).to.contain({ a: 'x', b: 'y' })
      expect(responseHeaders).to.contain({ a: 'x', b: 'y' })
    })

    it('gets headers values from an error response', async () => {
      let headers

      server.route({
        path: '/',
        method: 'GET',
        options: {
          handler: () => {
            throw Boom.unauthorized('Original message')
          },
          ext: {
            onPreResponse: {
              method: (request, h) => {
                request.response.output.headers.a = 'x'
                request.response.output.headers.b = 'y'

                headers = h.headers()

                return h.continue
              }
            }
          }
        }
      })

      const { headers: responseHeaders } = await server.inject('/')
      expect(headers).to.contain({ a: 'x', b: 'y' })
      expect(responseHeaders).to.contain({ a: 'x', b: 'y' })
    })
  })

  experiment('isView', () => {
    it('returns false when response is not available', async () => {
      server.route({
        path: '/',
        method: 'GET',
        options: {
          handler: (_, h) => {
            expect(h.isView()).to.be.false()

            return h.continue
          }
        }
      })

      await server.inject('/')
    })

    it('returns false when responding JSON', async () => {
      server.route({
        path: '/',
        method: 'GET',
        options: {
          handler: (_, h) => {
            expect(h.isView()).to.be.false()

            return { name: 'Marcus' }
          },
          ext: {
            onPreResponse: {
              method: (_, h) => {
                expect(h.isView()).to.be.false()

                return h.continue
              }
            }
          }
        }
      })

      await server.inject('/')
    })

    it('returns true when rendering a view', async () => {
      await server.register(Vision)

      server.views({
        engines: { hbs: require('handlebars') },
        path: Path.resolve(__dirname, 'fixtures/views')
      })

      server.route({
        path: '/',
        method: 'GET',
        options: {
          handler: async (_, h) => {
            return h.view('index')
          },
          ext: {
            onPreResponse: {
              method: (_, h) => {
                expect(h.isView()).to.be.true()
                return h.continue
              }
            }
          }
        }
      })

      const { result, statusCode } = await server.inject('/')
      expect(statusCode).to.equal(200)
      expect(result).to.include('<p>hapi-response-utilities are awesome</p>')
    })
  })
})


test/redirects.js
'use strict'

const Lab = require('@hapi/lab')
const Hapi = require('@hapi/hapi')
const { expect } = require('@hapi/code')

let server

const { describe, it, beforeEach } = (exports.lab = Lab.script())

describe('redirect responses:', () => {
  beforeEach(async () => {
    server = new Hapi.Server()
    await server.initialize()

    await server.register({
      plugin: require('../lib/index')
    })
  })

  describe('h.permanentRedirect ->', () => {
    it('redirects permanently with status 301', async () => {
      server.route({
        path: '/',
        method: 'GET',
        handler: (_, h) => h.permanentRedirect('/login')
      })

      const request = {
        url: '/',
        method: 'GET'
      }

      const response = await server.inject(request)
      expect(response.statusCode).to.equal(301)
    })

    it('redirects by default with status 302', async () => {
      server.route({
        path: '/',
        method: 'GET',
        handler: (_, h) => h.redirect('/login')
      })

      const request = {
        url: '/',
        method: 'GET'
      }

      const response = await server.inject(request)
      expect(response.statusCode).to.equal(302)
    })

    it('redirects temporarily with payload (status 307)', async () => {
      server.route([
        {
          method: 'POST',
          path: '/',
          handler: (_, h) => {
            return h.redirectWithPayload('/new')
          }
        }
      ])

      const request = {
        url: '/',
        method: 'POST'
      }

      const response = await server.inject(request)
      expect(response.statusCode).to.equal(307)
      expect(response.headers.location).to.equal('/new')
    })

    it('redirects permanently with payload (status 308)', async () => {
      server.route([
        {
          method: 'POST',
          path: '/',
          handler: (_, h) => {
            return h.permanentRedirectWithPayload('/new')
          }
        }
      ])

      const request = {
        url: '/',
        method: 'POST'
      }

      const response = await server.inject(request)
      expect(response.statusCode).to.equal(308)
      expect(response.headers.location).to.equal('/new')
    })
  })
})


test/status-shortcuts.js
'use strict'

const Lab = require('@hapi/lab')
const Hapi = require('@hapi/hapi')
const { expect } = require('@hapi/code')

let server

const { describe, it, beforeEach } = (exports.lab = Lab.script())

describe('shortcut methods for status codes:', () => {
  beforeEach(async () => {
    server = new Hapi.Server()
    await server.initialize()

    await server.register({
      plugin: require('../lib/index')
    })
  })

  describe('h.ok ->', () => {
    it('responds with status 200 and empty payload', async () => {
      server.route({
        path: '/',
        method: 'GET',
        handler: (_, h) => h.ok()
      })

      const request = {
        url: '/',
        method: 'GET'
      }

      const response = await server.inject(request)
      expect(response.statusCode).to.equal(200)
      expect(response.payload).to.be.empty()
    })

    it('responds with status 200 and payload', async () => {
      server.route({
        path: '/',
        method: 'GET',
        handler: (_, h) => h.ok({ works: 'fine' })
      })

      const request = {
        url: '/',
        method: 'GET'
      }

      const response = await server.inject(request)
      expect(response.statusCode).to.equal(200)
      expect(JSON.parse(response.payload)).to.equal({ works: 'fine' })
    })
  })

  describe('h.created ->', () => {
    it('responds with status 201 and empty payload', async () => {
      server.route({
        path: '/',
        method: 'GET',
        handler: (_, h) => h.created()
      })

      const request = {
        url: '/',
        method: 'GET'
      }

      const response = await server.inject(request)
      expect(response.statusCode).to.equal(201)
      expect(response.payload).to.be.empty()
    })

    it('responds with status 201 and payload', async () => {
      server.route({
        path: '/',
        method: 'GET',
        handler: (_, h) => h.created({ created: 'marcus' })
      })

      const request = {
        url: '/',
        method: 'GET'
      }

      const response = await server.inject(request)
      expect(response.statusCode).to.equal(201)
      expect(JSON.parse(response.payload)).to.equal({ created: 'marcus' })
    })
  })

  describe('h.accepted ->', () => {
    it('responds with status 202 and empty payload', async () => {
      server.route({
        path: '/',
        method: 'GET',
        handler: (_, h) => h.accepted()
      })

      const request = {
        url: '/',
        method: 'GET'
      }

      const response = await server.inject(request)
      expect(response.statusCode).to.equal(202)
      expect(response.payload).to.be.empty()
    })

    it('responds with status 202 and payload', async () => {
      server.route({
        path: '/',
        method: 'GET',
        handler: (_, h) => h.accepted({ link: 'futurestud.io' })
      })

      const request = {
        url: '/',
        method: 'GET'
      }

      const response = await server.inject(request)
      expect(response.statusCode).to.equal(202)
      expect(JSON.parse(response.payload)).to.equal({ link: 'futurestud.io' })
    })
  })

  describe('h.nonAuthoritativeInformation ->', () => {
    it('responds with status 203 and empty payload', async () => {
      server.route({
        path: '/',
        method: 'GET',
        handler: (_, h) => h.nonAuthoritativeInformation()
      })

      const request = {
        url: '/',
        method: 'GET'
      }

      const response = await server.inject(request)
      expect(response.statusCode).to.equal(203)
      expect(response.payload).to.be.empty()
    })

    it('responds with status 203 and payload', async () => {
      server.route({
        path: '/',
        method: 'GET',
        handler: (_, h) => h.nonAuthoritativeInformation({ like: 'woop-woop' })
      })

      const request = {
        url: '/',
        method: 'GET'
      }

      const response = await server.inject(request)
      expect(response.statusCode).to.equal(203)
      expect(JSON.parse(response.payload)).to.equal({ like: 'woop-woop' })
    })
  })

  describe('h.noContent ->', () => {
    it('responds with status 204 and ignores payload', async () => {
      server.route({
        path: '/',
        method: 'GET',
        handler: (_, h) => h.noContent({ ignores: 'payload' })
      })

      const request = {
        url: '/',
        method: 'GET'
      }

      const response = await server.inject(request)
      expect(response.statusCode).to.equal(204)
      expect(response.payload).to.be.empty()
    })
  })

  describe('h.resetContent ->', () => {
    it('responds with status 205 and ignores payload', async () => {
      server.route({
        path: '/',
        method: 'GET',
        handler: (_, h) => h.resetContent({ ignores: 'payload' })
      })

      const request = {
        url: '/',
        method: 'GET'
      }

      const response = await server.inject(request)
      expect(response.statusCode).to.equal(205)
      expect(response.payload).to.be.empty()
    })
  })

  describe('h.partialContent ->', () => {
    it('responds with status 206 and empty payload', async () => {
      server.route({
        path: '/',
        method: 'GET',
        handler: (_, h) => h.partialContent()
      })

      const request = {
        url: '/',
        method: 'GET'
      }

      const response = await server.inject(request)
      expect(response.statusCode).to.equal(206)
      expect(response.payload).to.be.empty()
    })

    it('responds with status 206 and payload', async () => {
      server.route({
        path: '/',
        method: 'GET',
        handler: (_, h) => h.partialContent({ partial: 'content' })
      })

      const request = {
        url: '/',
        method: 'GET'
      }

      const response = await server.inject(request)
      expect(response.statusCode).to.equal(206)
      expect(JSON.parse(response.payload)).to.equal({ partial: 'content' })
    })
  })
})


lib/index.js
'use strict'

async function register (server) {
  /**
   * Creates a PDF response forcing the browser to download the PDF file.
   *
   * @returns {Response}
   */
  server.decorate('toolkit', 'pdf', function (pdf, filename = 'download.pdf') {
    return this.response(pdf)
      .type('application/pdf')
      .header('Content-Description', 'File Transfer')
      .header('Content-Disposition', `attachment; filename="${filename}"`)
      .header('Content-Transfer-Encoding', 'binary')
  })

  /**
   * Sets a cookie by the given key-value-pair.
   *
   * @returns {Toolkit}
   */
  server.decorate('toolkit', 'cookie', function (key, value, options) {
    this.state(key, value, options)

    return this
  })

  /**
   * Sets a cookie by the given key-value-pair.
   *
   * @returns {Response}
   */
  server.decorate('toolkit', 'header', function (key, value, options) {
    const response = this.request.response

    if (!response) {
      throw new Error(`Cannot assign response headers [${key}/${value}] at this point, no response available yet (too early in the request lifecycle)`)
    }

    if (!response.isBoom) {
      response.header(key, value, options)

      return this
    }

    key = key.toLowerCase()
    const { headers } = response.output
    const { append = false, separator = ',', override = true, duplicate = true } = options || {}

    if ((!append && override) || !headers[key]) {
      headers[key] = value
    } else if (override) {
      if (key === 'set-cookie') {
        headers[key] = [].concat(headers[key], value)
      } else {
        if (!duplicate) {
          const values = headers[key].split(separator)

          for (const v of values) {
            if (v === value) {
              return this
            }
          }
        }

        headers[key] = headers[key] + separator + value
      }
    }

    return this
  })

  /**
   * Returns the response headers.
   *
   * @returns {Object}
   */
  server.decorate('toolkit', 'headers', function () {
    const response = this.request.response

    if (!response) {
      throw new Error('Cannot return response headers at this point, no response available yet (too early in the request lifecycle)')
    }

    return response.isBoom
      ? response.output.headers
      : response.headers
  })

  /**
   * Determine whether the response is a view.
   *
   * @returns {Boolean}
   */
  server.decorate('toolkit', 'isView', function () {
    const response = this.request.response || {}

    return response.variety === 'view'
  })

  /**
   * Redirect the given `path` with HTTP status 301.
   *
   * @returns {Response}
   */
  server.decorate('toolkit', 'permanentRedirect', function (path) {
    return this.redirect(path).code(301)
  })

  /**
   * Temporarily redirect the requested `path` with HTTP status 307.
   * Details: https://developer.mozilla.org/de/docs/Web/HTTP/Status
   *
   * @returns {Response}
   */
  server.decorate('toolkit', 'redirectWithPayload', function (path) {
    return this.redirect(path).code(307)
  })

  /**
   * Permanently redirect the requested `path` with HTTP status 308.
   * Details: https://developer.mozilla.org/de/docs/Web/HTTP/Status
   *
   * @returns {Response}
   */
  server.decorate('toolkit', 'permanentRedirectWithPayload', function (path) {
    return this.redirect(path).code(308)
  })

  /**
   * Shortcut to respond with just an HTTP status code.
   *
   * @returns {Response}
   */
  server.decorate('toolkit', 'status', function (code) {
    return this.response().code(code)
  })

  /**
   * Shortcut for a 200 OK response.
   *
   * @returns {Response}
   */
  server.decorate('toolkit', 'ok', function (payload) {
    return this.response(payload).code(200)
  })

  /**
   * Shortcut for a 201 Created response.
   *
   * @returns {Response}
   */
  server.decorate('toolkit', 'created', function (payload) {
    return this.response(payload).code(201)
  })

  /**
   * Shortcut for a 202 Accepted response.
   *
   * @returns {Response}
   */
  server.decorate('toolkit', 'accepted', function (payload) {
    return this.response(payload).code(202)
  })

  /**
   * Shortcut for a 203 Non-authoritative Information response.
   *
   * @returns {Response}
   */
  server.decorate('toolkit', 'nonAuthoritativeInformation', function (payload) {
    return this.response(payload).code(203)
  })

  /**
   * Shortcut for a 204 No Content response.
   *
   * @returns {Response}
   */
  server.decorate('toolkit', 'noContent', function () {
    return this.response().code(204)
  })

  /**
   * Shortcut for a 205 Reset Content response.
   *
   * @returns {Response}
   */
  server.decorate('toolkit', 'resetContent', function () {
    return this.response().code(205)
  })

  /**
   * Shortcut for a 206 Partial Content response.
   *
   * @returns {Response}
   */
  server.decorate('toolkit', 'partialContent', function (payload) {
    return this.response(payload).code(206)
  })
}

exports.plugin = {
  register,
  once: true,
  pkg: require('../package.json')
}
